gh-83371: handle exceptions from callbacks in process pools

duaneg opened the pull request:
User-supplied callbacks are called from an internal pool management thread. At present any exceptions they raise are not caught and so propagate out and kill the thread. This then causes problems for subsequent pool operations, including joining the pool hanging.

As a QoL improvement, catch and handle any such exceptions using the system exception hook. Thus by default details of the exception will be printed to stderr, but the pool's integrity will remain intact.

<!-- gh-issue-number: gh-83371 -->
* Issue: gh-83371
<!-- /gh-issue-number -->

duaneg committed:
gh-83371: handle exceptions from user-supplied callbacks in process pools

User-supplied callbacks are called from an internal pool management thread. At
present any exceptions they raise are not caught and so propagate out and kill
the thread. This then causes problems for subsequent pool operations, including
joining the pool hanging.

As a QoL improvement, catch and handle any such exceptions using the system
exception hook. Thus by default details of the exception will be printed to
stderr, but the pool's integrity will remain intact.

graingert commented on Lib/multiprocessing/pool.py:
```diff
@@ -776,13 +777,20 @@ def get(self, timeout=None):
+        try:
+            return callback(args)
+        except Exception as e:
+            sys.excepthook(*sys.exc_info())
```
maybe threading.excepthook would be better to call?

duaneg replied:
Yes, quite possibly. In fact I did use that in an earlier version of the code. I decided to change because the thread is internal to the pool code, which does not set its `excepthook`. So ISTM it will always end up using the system hook anyway, and the code is a little uglier because it needs to marshal the args then explicitly `del` them to avoid reference cycles:
```
    except Exception as e:
            args = threading.ExceptHookArgs([type(e), e, e.__traceback__, None])
            threading.excepthook(args)
            del args
```

However, it probably is more robust/future-proof/correct to use `threading.excepthook`, so I'm happy to switch to this if you think it is appropriate.

duaneg committed:
Use threading.excepthook instead of sys.excepthook

duaneg committed:
Raise new group exceptions on callback failure instead of just processing them
with the default exception hook.

This is a breaking change to the API, but only in cases where the existing code
would be completely broken anyway, so hopefully it isn't a problem.

TODO: docs need updating

duaneg committed:
Correct silly mistake

duaneg committed:
Move timing dependent test (of multiple callback failures) into separate
sub-test and use a barrier to ensure it is reliable. We must ensure all tasks
are submitted before any of them run and mark the pool as broken. Since
barriers can't be shared between processes, do not run that sub-test with
process-based parallelism.

duaneg committed:
CI tests on free-threading builds are failing due to leaked barrier resources.
Attempt to fix that by not creating the CallbackError inside the except clause
where the exceptions that cause them are caught, but just storing those
exceptions and creating the exception group later when required.

Also ensure if we are exiting a pool's context manager due to a BrokenPoolError
that we don't re-raise it with the same error added again.

duaneg committed:
Shared objects are being leaked via what *looks like* a reference cycle: try
harder to collect them. This doesn't seem like it should be necessary or make a
difference, but let's give it a try...

duaneg committed:
Merge branch 'main' of https://github.com/python/cpython into gh-83371

duaneg committed:
Revert all commits after "Use threading.excepthook instead of sys.excepthook"

This reverts this branch back to just calling `threading.excepthook`, instead
of trying to raise the exception when the user gets the result or exits the
context manager. I would like to get the immediate bug fixed, and I think this
is a simple, low-risk fix that does that.

From what I understand it isn't possible to prevent the reference cycle which
is causing all those test failures. We need to store the exception to raise it
later, which implicitly creates one.

We can always add raising the exception as an additional improvement later, but
in the meantime if any user code wants to handle the exceptions it can do so by
installing a default exception handler.

duaneg committed:
Add a test and a blurb

duaneg committed:
Correct reference in blurb
